code/Player/PropPossession.cs

```diff
@@ -0,0 +1,207 @@
+				// (Collection was modified; enumeration operation may not execute)
+				RemoveThisComponentDelayed();
+			}
+
+			if ( Host.IsClient && somePlayer.IsLocalPawn )
+				_nameplate?.Delete(); // local player has come alive, must loose all nameplates
+		}
+		else if ( somePlayer.Status != PlayerStatus.Alive )
+		{
+			// somePlayer has died
+			if ( somePlayer.IsLocalPawn )
+				_nameplate = new PossessionNameplate( _player, Prop ); // local player has died => show nameplates
+		}
+	}
+
+	private async void RemoveThisComponentDelayed()
+	{
+		await GameTask.Delay( 1 );
+		_player?.Components.RemoveAny<PropPossession>();
+	}
+
+	[Event.Tick.Server]
+	private void RechargePunchesAndCheckProp()
+	{
+		if ( !Prop.IsValid() )
+		{
+			RemoveThisComponentDelayed(); // delayed removal for the same reason as above
+			return;
+		}
+
+		if ( _timeUntilRecharge )
+		{
+			Punches = Math.Min( Punches + 1, Game.PropPossessionMaxPunches );
+			_timeUntilRecharge = Game.PropPossessionRechargeTime;
+		}
+	}
+
+	[Event.BuildInput]
+	private void BuildInput( InputBuilder input )
+	{
+		if ( !Entity.IsValid() || !Entity.IsLocalPawn )
+			return;
+
+		if ( input.Pressed( InputButton.Duck ) )
+		{
+			CancelPossession();
+			return;
+		}
+
+		float forward = input.Pressed( InputButton.Forward ) ? 1f : (input.Pressed( InputButton.Back ) ? -1f : 0f);
+		float left = input.Pressed( InputButton.Left ) ? 1f : (input.Pressed( InputButton.Right ) ? 1f : 0f);
+		bool up = input.Pressed( InputButton.Jump );
+		var rotation = Rotation.From( input.ViewAngles );
+
+		if ( forward + left != 0f || up )
+			MoveProp( forward, left, up, rotation );
+
+		input.SetButton( InputButton.Jump, false ); // cancel jump button so that spectator camera is not changed
+	}
+
+	[ConCmd.Server]
+	public static void BeginPossession( int propNetworkIdent )
+	{
+		if ( !Game.PropPossessionEnabled || ConsoleSystem.Caller.Pawn is not Player player )
+			return;
+
+		if ( player.IsAlive() || player.Components.TryGet<PropPossession>( out _ ) )
+			return;
+
+		var target = Sandbox.Entity.FindByIndex( propNetworkIdent );
+
+		if ( !target.IsValid() || target is not Prop prop || prop.PhysicsBody is null || target.Owner is Player )
+			return;
+
+		player.Components.Add( new PropPossession( prop ) );
+	}
+
+	[ConCmd.Server]
+	private static void CancelPossession()
+	{
+		ConsoleSystem.Caller.Pawn.Components.RemoveAny<PropPossession>();
+	}
+
+	[ConCmd.Server]
+	private static void MoveProp( float forward, float left, bool up, Rotation rotation )
+	{
+		ConsoleSystem.Caller.Pawn.Components.Get<PropPossession>().HandlePropMovement( forward, left, up, rotation );
+	}
+
+	private void HandlePropMovement( float forward, float left, bool up, Rotation rotation )
+	{
+		var b = Prop.PhysicsBody;
+
+		// reference:
+		// https://github.com/Facepunch/garrysmod/blob/master/garrysmod/gamemodes/terrortown/gamemode/propspec.lua#L111
+
+		if ( Punches <= 0 )
+			return;
+
+		if ( forward + left == 0f && !up )
```

Comment:
I think we do this check previously as well

Reply:
Yeah checking this serverside is not necessary since its checked clientside already.

I believe the clientside check is very much needed, so that the server is not continuously spammed with commands even when no button is pressed.

Another thing to keep in mind is that `forward` and `left` are currently used in `ApplyForceAt()` and `ApplyAngularImpulse()`. Im not sure if it is possible to call server commands like `MoveProp()` manually as a client.
Should a client manage to call `MoveProp(float forward, float left, ...)` manually then that client can decide how much force is applied.
Should the given `forward` and `left` values be checked for validity or do we not care?